### ISSUE TITLE
[FEATURE REQUEST]: document blob read push down?

### ISSUE THREAD

**Issue by Plasma** [dotnet/spark#368]
**Is your feature request related to a problem? Please describe.**
I’m wanting to understand how I tell spark which blobs under an azure blob container should be loaded, when the blobs are organised by date in sub folders.

I can’t spot any documentation on how to do this in a production scenario, any documentation or guidance appreciated.

**Comment by imback82** [comment]
@Plasma I need to understand your production scenario a bit more. Do you need to read all files under those sub folders, or do you plan to do partition pruning based on the date ?

**Comment by Plasma** [comment]
Partition pruning I think is what I meant.

As an example, assume I have a list of blobs like this in storage:

```
...
/my-container/logs/2019/11/30/log1.log
/my-container/logs/2019/11/30/log2.log
/my-container/logs/2019/12/01/log1.log
/my-container/logs/2019/12/01/log2.log
/my-container/logs/2019/12/01/log3.log
/my-container/logs/2019/12/02/log1.log
/my-container/logs/2019/12/03/log1.log
...
```

I can configure spark to read blobs in that folder structure like this:

```
path = 'wasb[s]://example.blob.core.windows.net/my-container/logs/'
spark.read.csv(path ).getOrCreateTempView("logs");
```

Then run a query:
```
val df = spark.sql("SELECT * FROM logs WHERE Timestamp >= '2019-12-01' AND Timestamp < '2019-12-03'
```

But how do I inform spark about being able to do partition pruning (I think?) where it won't read _all_ blobs in the `/my-container/logs/` folder, since it can instead just look at `/logs/2019-12-01` and `/logs/2019-12-02`.

Basically, with a huge number of logs in a folder, how can Spark be told to only need to look at a certain amount of logs based on the query?

I've assumed this should be a common production scenario but cannot find any examples of how to do this. If there's a better way please let me know.

I may not be phrasing the question correctly but I hope that makes sense -- thank you.

**Comment by imback82** [comment]
You can search online with `spark partition pruning` and find some answers like https://mapr.com/blog/tips-and-best-practices-to-take-advantage-of-spark-2-x/. I think you need `hive partitioning`; for example, you can write the data to folders like `year=2019`, `year=2018`, etc. and you can use `year` in your query to filter subfolders.

**Comment by Plasma** [comment]
Thanks for that. I'm guessing I can't use regular blob storage for this optimization?

If so, is it a bit redundant to then need to copy all data from blob storage to another storage system just for this feature to work?
